src/lavinmq/launcher.cr

```diff
@@ -27,28 +41,36 @@ module LavinMQ
       end
       Dir.mkdir_p @config.data_dir
       if @config.data_dir_lock?
-        @data_dir_lock = DataDirLock.new(@config.data_dir).tap &.acquire
+        @data_dir_lock = DataDirLock.new(@config.data_dir)
       end
-      @replicator = replicator = etcd ? Clustering::Server.new(@config, etcd) : Clustering::NoopServer.new
-      @amqp_server = LavinMQ::Server.new(@config, replicator)
-      @http_server = LavinMQ::HTTP::Server.new(@amqp_server)
+
+      if @config.clustering?
+        etcd = Etcd.new(@config.clustering_etcd_endpoints)
+        @runner = controller = Clustering::Controller.new(@config, etcd)
+        @replicator = Clustering::Server.new(@config, etcd, controller.id)
+      else
+        @replicator = Clustering::NoopServer.new
+        @runner = DefaultRunner.new
+      end
+
       @tls_context = create_tls_context if @config.tls_configured?
       reload_tls_context
       setup_signal_traps
-      setup_log_exchange
     end
 
-    def start : self
-      listen
+    private def start : self
+      @data_dir_lock.try &.acquire
```

Comment:
Does the replication client acquire the datadirlock independently?

Reply:
Yes, and I think that's good. I actually had a "bug" where the client didn't exit which I discovered when the server tried to startup after it was elected leader but couldn't aquire the lock.

Reply:
So the clustering client will modify the data dir without a lock? What about acquiring the lock in the launcher in the constructor?

Reply:
Both `Launcher` and `Clustering::Client` will lock, and it was thanks to this I noticed the bug that the `Client` didn't exit properly when the follower was elected leader, because `Launcher` couldn't aquire the lock.

It has always been likes this. (The bug was just during my refactoring, nothing that has been commited.)

Reply:
oh, i misread! yes, good that replication client also acquire :)